compiler/next/include/chpl/uast/Array.h

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2021 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef CHPL_UAST_ARRAY_H
+#define CHPL_UAST_ARRAY_H
+
+#include "chpl/queries/Location.h"
+#include "chpl/uast/Expression.h"
+
+namespace chpl {
+namespace uast {
+
+
+/**
+  This class represents an array expression. For example:
+
+  \rst
+  .. code-block:: chapel
+
+      // Example 1:
+      var a = [ 1, 2, 3 ];
+
+  \endrst
+
+ */
+class Array final : public Expression {
+ private:
+  // TODO: Record if initializer list has trailing comma?
+  Array(ASTList children)
+    : Expression(asttags::Array, std::move(children)) {
+    assert(isExpressionASTList(children_));
+  }
+
+  bool contentsMatchInner(const ASTNode* other) const override {
+    return expressionContentsMatchInner(other->toExpression());
+  }
+
+  void markUniqueStringsInner(Context* context) const override {
+    expressionMarkUniqueStringsInner(context);
+  }
+
+ public:
+  ~Array() override = default;
+
+  /**
+   Create and return an Array expression.
+   */
+  static owned<Array> build(Builder* builder, Location loc,
+                            ASTList exprs);
+
+  /**
+    Return a way to iterate over the expressions of this array.
+  */
+  ASTListIteratorPair<Expression> exprs() const {
```

Comment:
`elements` might be a reasonable alternative for the name

Reply:
Let's make it clear in the comment whether or not these could include Comments

Reply:
Do we want it to include comments?

Reply:
no

Reply:
Do we want to rename to `elements` if we're just keeping Domain as `exprs`?

Reply:
I'm not so sure. That's why I introduce it as a "reasonable alternative".

In particular, for an associative array, we might be parsing it as key1, val1, key2, val2 etc and then they're not all elements either. But I don't think we support that yet.